server/src/main/java/org/apache/iotdb/db/qp/strategy/LogicalGenerator.java

```diff
@@ -800,60 +815,87 @@ private TSDataType parseTypeNode(AstNode typeNode) throws LogicalOperatorExcepti
     if (timeInterval <= 0) {
       throw new LogicalOperatorException("Interval must more than 0.");
     }
-    String granu = node.getChild(1).getText();
-    switch (granu) {
-      case "w":
-        timeInterval *= 7;
-      case "d":
-        timeInterval *= 24;
-      case "h":
-        timeInterval *= 60;
-      case "m":
-        timeInterval *= 60;
-      case "s":
-        timeInterval *= 1000;
-      default:
-        break;
-    }
     return timeInterval;
   }
 
   private Pair<Path, String> parseLeafNode(AstNode node) throws LogicalOperatorException {
     if (node.getChildCount() != 2) {
       throw new LogicalOperatorException(
-              "error format in SQL statement, please check whether SQL statement is correct.");
+          "error format in SQL statement, please check whether SQL statement is correct.");
     }
     AstNode col = node.getChild(0);
     if (col.getType() != TqlParser.TOK_PATH) {
       throw new LogicalOperatorException(
-              "error format in SQL statement, please check whether SQL statement is correct.");
+          "error format in SQL statement, please check whether SQL statement is correct.");
     }
     Path seriesPath = parsePath(col);
     AstNode rightKey = node.getChild(1);
     String seriesValue;
-    if(rightKey.getChild(0).getType() == TqlParser.TOK_DATETIME){
+    if (rightKey.getChild(0).getType() == TqlParser.TOK_DATETIME) {
       if (!seriesPath.equals(SQLConstant.RESERVED_TIME)) {
         throw new LogicalOperatorException("Date can only be used to time");
       }
-      seriesValue = parseTokenTime(rightKey.getChild(0));
-    }else{
+      seriesValue = parseTokenTime(rightKey.getChild(0)) + "";
+    } else if (rightKey.getType() == TqlParser.TOK_DATE_EXPR) {
+      seriesValue = parseTokenDataExpresssion(rightKey.getChild(0)) + "";
+    } else {
       seriesValue = cascadeChildrenText(rightKey);
     }
     return new Pair<>(seriesPath, seriesValue);
   }
 
-  private String parseTokenTime(AstNode astNode) throws LogicalOperatorException {
-    return parseTimeFormat(cascadeChildrenText(astNode)) + "";
+  private Long parseTokenDataExpresssion(AstNode astNode) throws LogicalOperatorException {
+    if (astNode.getType() == TqlParser.PLUS) {
+      return parseTokenDataExpresssion(astNode.getChild(0)) + parseTokenDataExpresssion(
+          astNode.getChild(1));
+    } else if (astNode.getType() == TqlParser.MINUS) {
+      return parseTokenDataExpresssion(astNode.getChild(0)) - parseTokenDataExpresssion(
+          astNode.getChild(1));
+    } else {
+      return parseTokenTime(astNode);
+    }
+  }
+
+  private Long parseTokenTime(AstNode astNode) throws LogicalOperatorException {
+    if (astNode.getType() == TqlParser.TOK_DURATION) {
+      return parseTokenDuration(astNode);
+    }
+    return parseTimeFormat(cascadeChildrenText((astNode)));
+  }
+
+  private Long parseTokenDuration(AstNode astNode) {
+    String durationStr = cascadeChildrenText(astNode);
```

Comment:
it is better to add some comments to introduce what the AstNode looks like.. to help people understand the codes.

Reply:
ok, I'll add comments .